litr/src/main/java/com/linkedin/android/litr/io/Camera2MediaSource.kt

```diff
@@ -0,0 +1,253 @@
+package com.linkedin.android.litr.io
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.Rect
+import android.hardware.camera2.*
+import android.os.Build
+import android.os.Handler
+import android.os.HandlerThread
+import android.util.Log
+import android.util.Range
+import android.util.Size
+import android.view.Surface
+import android.view.SurfaceHolder
+import androidx.annotation.RequiresApi
+
+private const val TAG = "Camera2MediaSource"
+
+private const val DEFAULT_RECORD_WIDTH = 1280
+
+/**
+ * An implementation of MediaSource, which utilises Android's Camera2 APIs to capture from a
+ * CameraDevice and render to the ExternalMediaSource's input Surface. An instance of this class is
+ * expected to be both the MediaSource, and Decoder for which the pipeline is configured. This
+ * allows these components to be bypassed.
+ */
+@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+class Camera2MediaSource(
+        context: Context,
+        private val cameraId: String
+): ExternalMediaSource(), ExternalMediaSource.Callback {
+    /**
+     * Callback which notifies when the device is ready, as well as other details regarding the
+     * capture session.
+     */
+    interface Callback {
+        fun onDeviceReady(cameraCharacteristics: CameraCharacteristics)
+        fun onFrameSkipped(frameSkipCount: Int)
+
+        fun onError(exception: Exception)
+    }
+
+    var recordWidth = DEFAULT_RECORD_WIDTH
+
+    private var callback: Callback? = null
+    private var inputSurface: Surface? = null
+
+    // We support an optional SurfaceHolder that can be used as a Preview for the Camera capture.
+    private var previewSurfaceHolder: SurfaceHolder? = null
+    private val previewSurfaceHolderListener = object : SurfaceHolder.Callback {
+        override fun surfaceCreated(holder: SurfaceHolder) {
+            Log.i(TAG, "Preview Surface Created")
+            previewSurfaceHolder = holder
+        }
+
+        override fun surfaceChanged(surfaceHolder: SurfaceHolder, p1: Int, p2: Int, p3: Int) {}
```

Comment:
`p1`, etc are not good parameter names, let's make them more descriptive

Reply:
Nice catch, thanks